register-machines/src/main.rs

```diff
@@ -11,55 +11,203 @@ enum Instruction {
+        }
+    }
+}
+
+impl<T: Encodable> Encodable for [T] {
+    fn encode(&self) -> Godel {
+        self.iter()
+            .rev()
+            .fold(
+                (None as Option<u128>).encode(),
+                |x, y| Some((y, x)).encode()
+            )
+    }
+}
+
+trait Decodable {
+    fn decode(godel: Godel) -> Self;
+}
+
+impl Decodable for u128 {
+    fn decode(godel: Godel) -> Self {
+        godel
+    }
+}
+
+impl Decodable for usize {
+    fn decode(godel: Godel) -> Self {
+        godel as usize
+    }
 }
+
+impl<T: Decodable, V: Decodable> Decodable for (T, V) {
+    fn decode(godel: Godel) -> Self {
+        let x_godel = godel.trailing_ones() as Godel;
+        let y_godel = godel >> (x_godel + 1);
+
+        (decode(x_godel), decode(y_godel))
+    }
+}
+
+impl<T: Decodable> Decodable for Option<T> {
+    fn decode(godel: Godel) -> Self {
+        if godel == 0 {
+            None
+        } else {
+            Some(decode(godel - 1))
+        }
+    }
+}
+
+impl Decodable for Instruction {
+    fn decode(godel: Godel) -> Self {
+        match decode::<Option<(u128, Godel)>>(godel) {
+            None => Halt,
+            Some((f, arg)) => {
+                if f % 2 == 0 {
+                    // EVEN: add
+                    Add(f / 2, decode(arg))
+                } else {
+                    // ODD: sub
+                    let (j, k) = decode(arg);
+                    Sub((f - 1) / 2, j, k)
+                }
+            }
+        }
+    }
+}
+
+impl<T: Decodable> Decodable for Vec<T> {
+    fn decode(mut godel: Godel) -> Self {
+        let mut list = vec!();
+
+        while let Some((head, tail)) = decode(godel) {
+            list.push(head);
+
+            godel = tail;
+        }
+    
+        list
+    }
+}
+
+fn decode<T: Decodable>(godel: Godel) -> T {
+    T::decode(godel)
+}
+
+use Instruction::*;
+
+fn eval_program(program: &[Instruction], init: &State) -> State {
+    let (mut label, mut registers) = init.clone();
+
+    while label < program.len() {
+        /* Reveal effects of instruction to state. */
+        match program[label] {
+            Add(reg, l) => {
+                /* Increment target register. */
+                *registers.get_mut(&reg).unwrap() += 1;
+    
+                /* Jump to label. */
+                label = l;
+            },
+            Sub(reg, lsub, lnop) => {
+                let register = registers.get_mut(&reg).unwrap();
+    
+                if *register == 0 {
+                    /* Register can't be decremented. */
+                    label = lnop;
+                } else {
+                    /* Decrement register. */
+                    *register -= 1;
+                    label = lsub;
+                }
+            },
+            Halt => { break; },
+        };
+    }
+
+    return (label, registers);
+}
+
+fn encode_program_to_list(program: &[Instruction]) -> Vec<u128> {
+    program.iter().map(Encodable::encode).collect()
+}
```

Comment:
I was half-expecting this code to produce some awful assembly when I first thought about how the compiler might handle type inference:
```
example::encode_program_to_list:
        push    r15
        push    r14
        push    r13
        push    r12
        push    rbx
        mov     r15, rdx
        mov     r13, rsi
        mov     r14, rdi
        test    rdx, rdx
        je      .LBB10_1
        mov     r12, r15
        shl     r12, 4
        mov     esi, 8
        mov     rdi, r12
        call    qword ptr [rip + __rust_alloc@GOTPCREL]
        test    rax, rax
        jne     .LBB10_3
        mov     esi, 8
        mov     rdi, r12
        call    qword ptr [rip + alloc::alloc::handle_alloc_error@GOTPCREL]
        ud2
.LBB10_1:
        mov     eax, 8
.LBB10_3:
        mov     qword ptr [r14], rax
        mov     qword ptr [r14 + 8], r15
        mov     qword ptr [r14 + 16], 0
        test    r15, r15
        je      .LBB10_4
        lea     r8, [8*r15 - 8]
        shl     r15, 3
        shr     r8, 3
        add     r8, 1
        lea     r10, [r15 + 4*r15]
        xor     r9d, r9d
        xor     esi, esi
        jmp     .LBB10_6
.LBB10_12:
        movzx   ecx, byte ptr [r13 + rsi + 8]
        add     cl, cl
        mov     rdi, qword ptr [r13 + rsi + 24]
        lea     rdx, [rdi + rdi]
        add     rdx, 1
        shr     rdi, 63
        shld    rdi, rdx, cl
        shl     rdx, cl
        test    cl, 64
        cmovne  rdi, rdx
        mov     ebx, 0
        cmove   rbx, rdx
.LBB10_9:
        mov     qword ptr [rax], rbx
        mov     qword ptr [rax + 8], rdi
        add     rax, 16
        add     rsi, 40
        cmp     r10, rsi
        je      .LBB10_10
.LBB10_6:
        mov     rcx, qword ptr [r13 + rsi]
        test    rcx, rcx
        je      .LBB10_12
        mov     ebx, 0
        mov     edi, 0
        cmp     ecx, 1
        jne     .LBB10_9
        movzx   edx, byte ptr [r13 + rsi + 8]
        movzx   ecx, byte ptr [r13 + rsi + 24]
        add     dl, dl
        mov     rdi, qword ptr [r13 + rsi + 32]
        lea     rbx, [rdi + rdi]
        add     rbx, 1
        shr     rdi, 63
        shld    rdi, rbx, cl
        shl     rbx, cl
        test    cl, 64
        cmovne  rdi, rbx
        cmovne  rbx, r9
        shld    rdi, rbx, 1
        add     rbx, rbx
        add     rbx, -1
        adc     rdi, -1
        and     dl, 126
        or      dl, 1
        mov     ecx, edx
        shld    rdi, rbx, cl
        shl     rbx, cl
        test    dl, 64
        cmovne  rdi, rbx
        cmovne  rbx, r9
        jmp     .LBB10_9
.LBB10_4:
        xor     r8d, r8d
.LBB10_10:
        mov     qword ptr [r14 + 16], r8
        mov     rax, r14
        pop     rbx
        pop     r12
        pop     r13
        pop     r14
        pop     r15
        ret
```
It's entirely inlined. The only `call` here is to allocate the return vector for `.collect()`, which happens only once because the size of the return vec is known at the time the function is called. I simply love how efficient this code is.